### PR TITLE
Added authentication

### PR DESCRIPTION
**Pull request by muskanbagrecha**
- Added auth functionality - signup and login

- Users cannot add a product to the cart without logging in. On signup, the user does not have to login again. 
- I have a separate route for login but I am also using a modal that will be triggered if the user tries to add a product to the cart without logging in. 
- You can view the login page by clicking on the user icon on the navbar. The modal pops up when add to cart button is clicked on any of the product.

[Files to view](https://github.com/muskanbagrecha/ecommerce-react/pull/4/files?file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false)

PS: Wishlist button is not working yet.

### CODE REVIEW

File: src/Pages/Signuppage/Signupform.jsx [hunk 1]
```diff
@@ -0,0 +1,213 @@
+import { useState, useEffect } from "react";
+import { Eye } from "../../Assets/Icons/icons";
+import { Link, Navigate, useLocation, useNavigate } from "react-router-dom";
+import { useAuth } from "../../CustomHooks/useAuth";
+import { useModal } from "../../CustomHooks/useModal";
+import { useAlert } from "../../CustomHooks/useAlert";
```
Comment (zoef-yt) [inline src/Pages/Signuppage/Signupform.jsx]: Use index.js for this.

File: src/Pages/Signuppage/Signupform.jsx [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+      const response = await axios.post("/api/auth/signup", {
+        firstName,
+        lastName,
+        email,
+        password,
+      });
+      if (response.status === 201) {
+        setSignupDetails(intialState);
+        authDispatch({ type: "SIGNUP", payload: response.data });
+        // localStorage.setItem("user", JSON.stringify(response.data));
```
Comment (zoef-yt) [inline src/Pages/Signuppage/Signupform.jsx]: remove the commented code.

File: src/Pages/Signuppage/Signupform.jsx [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    });
+  };
+
+  const submitHandler = async (event) => {
+    event.preventDefault();
+    const { firstName, lastName, email, password, confirmPassword } =
+      signupDetails;
+    setError(initialErrorState);
+
+    if (password.length < 6) {
```
Comment (zoef-yt) [inline src/Pages/Signuppage/Signupform.jsx:56]: You can check if the password is alphanumeric.

File: src/Pages/Signuppage/Signupform.jsx [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+          First Name<span className="red">*</span>
+        </label>
+        <input
+          type="text"
+          name="firstName"
+          placeholder="John"
+          required
+          value={signupDetails.firstName}
+          onChange={signupInputChangeHandler}
+        />
```
Comment (zoef-yt) [inline src/Pages/Signuppage/Signupform.jsx:159]: As this elements are repeating you can just make it as one component

File: src/Pages/Loginpage/LoginModal.jsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import LoginForm from "./LoginForm";
+import { Modal } from "../../Components/UI/";
+import { useModal } from "../../CustomHooks/useModal";
```
Comment (zoef-yt) [inline src/Pages/Loginpage/LoginModal.jsx]: use index.js

File: src/Pages/Loginpage/LoginForm.jsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
+import { useState, useEffect } from "react";
+import { Link } from "react-router-dom";
+import { Eye } from "../../Assets/Icons/icons";
+import { useAuth } from "../../CustomHooks/useAuth";
+import { useFetch } from "../../CustomHooks/useFetch";
+import { useModal } from "../../CustomHooks/useModal";
```
Comment (zoef-yt) [inline src/Pages/Loginpage/LoginForm.jsx]: Use  index.js here.

File: src/Pages/Loginpage/LoginForm.jsx [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        <div className="login__CTA">
+          <button className="btn btn-primary">Login</button>
+          <Link className="btn btn-link-text align-center" to="/signup">
+            Create new account
+          </Link>
+          <button className="btn btn-outline" onClick={testLogin}>
+            Test Credentials
+          </button>
+        </div>
+        <span className="red error-message">{error}</span>
```
Comment (zoef-yt) [inline src/Pages/Loginpage/LoginForm.jsx]: You should add a check for error span to show or not/

File: src/Context/modalContext.jsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import { createContext } from "react";
+import { useState } from "react";
```
Comment (zoef-yt) [inline src/Context/modalContext.jsx]: Make it in one line

File: src/Context/authContext.jsx [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  //       setError("Invalid email or password. Please try again.");
+  //     } else {
+  //       setError("Unexpected error. Please try again in some time.");
+  //     }
+  //   }
+  // };
+
+  // const logoutHandler = () => {
+  //   authDispatch({ type: "LOGOUT" });
+  // };
```
Comment (zoef-yt) [inline src/Context/authContext.jsx]: Please remove the unused code.

File: src/Context/authContext.jsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+const AuthContext = createContext({
+  isAuthenticated: false,
+  token: null,
+  user: {},
+});
+
+const AuthProvider = ({ children }) => {
+  // const [loader, setLoader] = useState(false);
+  // const [error, setError] = useState(null);
```
Comment (zoef-yt) [inline src/Context/authContext.jsx]: remove the commented code

File: src/Context/authContext.jsx [hunk 11]
```diff
@@ -0,0 +1,66 @@
+import { createContext } from "react";
+import { useFetch } from "../CustomHooks/useFetch";
+import { authReducer } from "../Reducer/authReducer";
+import { useReducer, useState } from "react";
```
Comment (zoef-yt) [inline src/Context/authContext.jsx]: use index.js

File: src/Components/UI/index.jsx [hunk 12]
```diff
@@ -2,7 +2,8 @@ import Footer from "./Footer/Footer";
 import Navigation from "./Navigation/Navigation";
 import Carousel from "./Carousel/Carousel";
 import Card from "./Card/Card";
-import Pagination from "./Pagination/Pagination";
 import Section from "./Section/Section";
+import Alert from "./Alert/Alert";
+import Modal from "./Modal/Modal";
```
Comment (zoef-yt) [inline src/Components/UI/index.jsx:7]: index.js

File: src/Components/UI/Navigation/Navigation.jsx [hunk 13]
```diff
@@ -57,19 +68,40 @@ const Navigation = () => {
...
+            )}
+            {!isAuthenticated && (
+              <div>
+                <p>Welcome,</p>
+                <small>To access account</small>
+                <Link to="/login" className="btn-link-text">
+                  Login/Signup
+                </Link>
+              </div>
+            )}
```
Comment (zoef-yt) [inline src/Components/UI/Navigation/Navigation.jsx]: Use a ternary operator here.

File: src/Components/UI/Navigation/Navigation.jsx [hunk 14]
```diff
@@ -57,19 +68,40 @@ const Navigation = () => {
...
-              <User />
-            </i>
+          <NavLink to={isAuthenticated ? "/user" : "/login"}>
+            {!isAuthenticated ? (
+              <i className="btn-icon">
+                <User />
+              </i>
+            ) : (
+              <span className="avatar avatar-xs avatar-text">
+                {authState.user.firstName[0] + authState.user.lastName[0]}
```
Comment (zoef-yt) [inline src/Components/UI/Navigation/Navigation.jsx:77]: Appreciation: Great work here to show two initials of the user name.

File: src/Components/UI/Navigation/Navigation.jsx [hunk 15]
```diff
@@ -1,21 +1,24 @@
 import React from "react";
 import { useState } from "react";
-import { Link, NavLink, useNavigate } from "react-router-dom";
+import { Link, NavLink, useNavigate, useLocation } from "react-router-dom";
 import { Cart, User, Wishlist, Search } from "../../../Assets/Icons/icons";
 import { useFilter } from "../../../CustomHooks/useFilter";
+import { useAuth } from "../../../CustomHooks/useAuth";
```
Comment (zoef-yt) [inline src/Components/UI/Navigation/Navigation.jsx]: index.js

**Review by zoef-yt — CHANGES_REQUESTED** [review]

File: src/Pages/Signuppage/Signupform.jsx (hunk 4 above)
Comment (muskanbagrecha) [reply src/Pages/Signuppage/Signupform.jsx:159]: Sure. I will do that at the end while refactoring the code.

File: src/Components/UI/index.jsx (hunk 12 above)
Comment (muskanbagrecha) [reply src/Components/UI/index.jsx:7]: This is index.jsx xD

Comment (zoef-yt) [reply src/Components/UI/index.jsx:7]: My bad

File: src/Pages/Signuppage/Signupform.jsx (hunk 4 above)
Comment (zoef-yt) [reply src/Pages/Signuppage/Signupform.jsx:159]: Okay

**Review by zoef-yt — APPROVED** [review]
The code looks really good. The UI looks absolutely amazing.
UI Appreciation: Really loved your toast on auth page. Great work muskan. You are good to merge.

**Review by sahilsharma1901 — APPROVED** [review]

**Comment by sahilsharma1901** [comment]
Code looks good to me, especially the use of custom hooks. In the UI, the layout with proper set of images makes it more elegant. 

Also, the auth flow is seamless.

**Pull request merged**